src/lib/reasoners/satml_frontend.ml

```diff
@@ -1035,6 +1038,110 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
+     exploration of this branch of the SAT solver. It doesn't mean there is no
+     other branch of the SAT solver in which all the formulas are satisfiable
+     and the objective functions have bigger values in some models of this
+     branch.
+
+     For instance, consider the SMT-LIB problem:
+       (declare-const x Int)
+       (assert (or (<= x 2) (<= x 4)))
+       (assert (or (<= y 3))
+       (maximize x)
+       (maximize y)
+       (check-sat)
+       (get-model)
+
+     Assume that the solver explores the first branch (<= x 2) of the or gate:
+       (or (<= x 2) (<= x 4)).
+
+     Let's imagine it discovers that [0] is a model of the first formula. After
+     optimization, it find that [2] is the best model for [x] and [3] is the
+     best model for [y] it can got in this branch. Still we have to explore
+     the second branch [(<= x 4)] to realize that [4] is actually the best
+     model for [x].
+
+     The following function ensures to explore adequate branches of the SAT
+     solver in order to get the best model, if the problem is SAT.
+
+     In our running example, after getting the model [2], the below function
+     run again the SAT solver with the extra assert:
+       (assert (or (> x 2) (and (= x 2) (> y 3)))
+
+     If this run produces the answer [unsat], we know that [2] was the best
+     model value for [x]. Otherwise, we got a better value for [x]. *)
+
+  let op is_max =
+    match is_max with
+    | true -> Expr.Reals.(<=)
+    | false -> Expr.Reals.(>=)
+
+  let analyze_unknown_for_objectives env unknown_exn unsat_rec_prem : unit =
+    let obj = Th.get_objectives (SAT.current_tbox env.satml) in
+    if Util.MI.is_empty obj then raise unknown_exn;
+    env.objectives := Some obj;
+    let acc =
+      try
+        Util.MI.fold
+          (fun _ {Th_util.e; value; r=_; is_max; _} acc ->
+             match value with
+             | Pinfinity | Minfinity ->
+               raise (Give_up acc)
+             | Value (v, epsilon) ->
+               begin
+                 match epsilon with
+                 | Plus | Minus -> raise (Give_up ((e, v, is_max) :: acc))
+                 | None -> (e, v, is_max) :: acc
+               end
+             | Unknown ->
+               assert false
+          ) obj []
+      with Give_up acc -> acc
+    in
+    begin match acc with
+      | [] ->
+        (* The answer for the first objective is infinity. We stop here as
+           we cannot go beyond infinity and the next objectives with lower
+           priority cannot be optimized in presence of infinity values. *)
+        raise unknown_exn;
+
+      | (e, tv, is_max) :: l ->
+        let neg =
+          List.fold_left
+            (fun acc (e, tv, is_max) ->
+               let eq = E.mk_eq e tv ~iff: false in
+               let acc = E.Core.and_ acc eq in
+               E.Core.(or_ acc (not ((op is_max) e tv)))
+            ) (E.Core.not ((op is_max) e tv)) l
```

Comment:
Actually, I knew the code was suspicious but I didn't manage to trigger the bug. I thought replacing the new constraint `x > 2.0` by `x >= 2.0` could produce an infinity loop. I guess there is no infinity loop because the constraint `x >= 2.0` and `x <= 2.0` will produce the equation `x = 2.0` and the `optimizing_split` of `IntervalCalculus` produces an equation case-split in this case.

Reply:
> I didn't manage to trigger the bug

To be fair it was a bit tricky to manage to do so :) In general however it is good to mention it (either as a comment in the code or on the PR) when you know that some code is suspicious or you are unsure it is correct, it hints at the reviewer to do a more careful review of that part of the code and helps reduce (but not eliminate, sadly) bugs.